### ISSUE TITLE
Media query class hashes aren't guaranteed to be unique

### ISSUE THREAD

**Issue by tptee** [FormidableLabs/radium#593]
Steps to reproduce:
- Clone [radium-grid](https://github.com/FormidableLabs/radium-grid) and `npm install`.
- Run `builder run server-hot`.
- View the demo at localhost:3000.
- View the "Custom per-cell widths" grid.

Expected results:
Screens in order from small -> medium -> large -> xlarge.
<img width="346" alt="screenshot 2016-02-16 10 30 46" src="https://cloud.githubusercontent.com/assets/4156258/13086628/e1c30100-d498-11e5-9702-e81d138ec3bd.png">
<img width="607" alt="screenshot 2016-02-16 10 30 55" src="https://cloud.githubusercontent.com/assets/4156258/13086635/e772446c-d498-11e5-93aa-66190babe957.png">
<img width="880" alt="screenshot 2016-02-16 10 31 08" src="https://cloud.githubusercontent.com/assets/4156258/13086639/ed24faf8-d498-11e5-9d7e-bfb763dd2d77.png">
<img width="1041" alt="screenshot 2016-02-16 10 31 19" src="https://cloud.githubusercontent.com/assets/4156258/13086645/f157e414-d498-11e5-91b7-46b9914e3c85.png">

Actual results:
Screens in same order as above.
<img width="357" alt="screenshot 2016-02-16 10 33 31" src="https://cloud.githubusercontent.com/assets/4156258/13086655/ffa36e3a-d498-11e5-91fe-1ebeb4c78fae.png">
<img width="735" alt="screenshot 2016-02-16 10 33 22" src="https://cloud.githubusercontent.com/assets/4156258/13086660/0499df96-d499-11e5-955d-3c20efb0d4dc.png">
<img width="1048" alt="screenshot 2016-02-16 10 33 15" src="https://cloud.githubusercontent.com/assets/4156258/13086661/08911146-d499-11e5-8cb9-16c609ce6f52.png">
<img width="1040" alt="screenshot 2016-02-16 10 33 04" src="https://cloud.githubusercontent.com/assets/4156258/13086664/0c412948-d499-11e5-8cbf-18ee67def1f6.png">

The CSS for a working grid looks like this:
<img width="562" alt="screenshot 2016-02-16 10 37 11" src="https://cloud.githubusercontent.com/assets/4156258/13086740/5b6bf3c2-d499-11e5-99d6-1b3fe9aada92.png">

The CSS for the broken grid has gaps where styles disappear for some breakpoints (notice the missing 641):
<img width="560" alt="screenshot 2016-02-16 10 37 28" src="https://cloud.githubusercontent.com/assets/4156258/13086764/78339b68-d499-11e5-90db-1def5f3adf5a.png">

It appears that using the media query and its styles does not guarantee uniqueness, since multiple elements can have identical media query/style combos. Therefore, StyleKeeper considers some styles as duplicates, never entering them into the cssSet.

Changing the hash to a UUID delivers the expected results without breaking any tests. Would you like a PR for this, or is there a better source of uniqueness we can derive?

**Comment by tptee** [comment]
An alternative to a UUID could be either the key or id of the style's owning component?

**Comment by ianobermiller** [comment]
Yes, hashing the content is a problem for just this reason. It becomes even
worse when you consider using StyleKeeper for :hover and friends. The only
way a UUID will work is if we can reproducibly seed it. We can probably
keep the generator on StyleKeeper -- it must be reproducible for
server-side rendering to work correctly.

**Comment by tptee** [comment]
What do you think of the following?

New code:
- Tiny `styleId` plugin, run after mergeStyleArray, that adds a unique ID prop to all enhanced components (unless they already have one):
```es6
({ props }) => {
  if (!props.styleId) {
    return {
      props: {styleId: uuid()}
    };
  }
},
```
- StyleKeeper plugins like resolveMediaQueries use props.styleId instead of the old hash:
```es6
const mediaQueryClassName = 'rmq-' + props.styleId;
```

Server render:
1) Adds `styleId` to the props of all enhanced components.
2) addCSS uses the component's `styleId` prop.

Client:
1) Initially nothing. The server generated the components' `styleId` props.
2) If a new component enters on the client, it's given a `styleId` prop and addCSS'd.

Will this be enough to keep the server render in sync with the client?

**Comment by ianobermiller** [comment]
@tptee that seems like a lot of work; why not just have an incrementing counter in `StyleKeeper`? Should work the same way.
